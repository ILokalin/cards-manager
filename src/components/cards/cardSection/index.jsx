import { getCardType, CardFactory, emptyCardTemplate } from "utils/card";
import { CardNavBar } from "components/cards/cardNavBar";

export const CardSection = ({ card = emptyCardTemplate }) => {
  return (
    <div className="pt-4">
      <h2 className="mb-3">Карточка&nbsp;{getCardType(card).genitive}</h2>
      {CardFactory[card.type](card)}
      <CardNavBar />
    </div>
  );
};
