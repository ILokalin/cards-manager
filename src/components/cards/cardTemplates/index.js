export { CardStudent } from "./cardStudent";
export { CardCover } from "./cardCover";
