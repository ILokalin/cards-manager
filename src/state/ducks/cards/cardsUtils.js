import { sortAZ, sortZA } from "utils/sort";

const sortSelector = (isAZDirection) => {
  return [sortZA, sortAZ][Number(isAZDirection)];
};

export const setSortState = (state, payload) => {
  let { isAZDirection } = state.sort;
  if (state.sort.key === payload) {
    isAZDirection = !isAZDirection;
  } else {
    isAZDirection = true;
  }

  return state
    .set("cards", [...state.cards.sort(sortSelector(isAZDirection)(payload))])
    .set("sort", { key: payload, isAZDirection });
};

export const selectCardState = (state, payload) => {
  return state.set(
    "cards",
    state.cards.map((card) => {
      return { ...card, isActive: card.id === payload ? true : false };
    })
  );
};

export const navForwardState = (state) => {
  const { length } = state.cards;
  const currentIndex = state.cards.findIndex((card) => card.isActive);
  const newIndex = currentIndex === length - 1 ? 0 : currentIndex + 1;

  return state.set(
    "cards",
    state.cards.map((card, indx) => {
      return { ...card, isActive: indx === newIndex ? true : false };
    })
  );
};

export const navBackwardState = (state) => {
  const { length } = state.cards;
  const currentIndex = state.cards.findIndex((card) => card.isActive);
  const newIndex = currentIndex === 0 ? length - 1 : currentIndex - 1;

  return state.set(
    "cards",
    state.cards.map((card, indx) => {
      return { ...card, isActive: indx === newIndex ? true : false };
    })
  );
};

export const getCardsState = (state, payload) => {
  payload.sort(sortSelector(state.sort.isAZDirection)(state.sort.key));
  payload[0].isActive = true;

  return state.set("cards", [...payload]).set("isLoading", false);
};

export const startLoadingState = (state) => state.set("isLoading", true);
