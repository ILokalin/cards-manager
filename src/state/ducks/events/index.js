import { Record } from "immutable";
import { useSelector } from "react-redux";

export const moduleName = "events";
const prefix = `${moduleName}`;

export const SHOW_ALERT = `${prefix}/SHOW_ALERT`;
export const REMOVE_ALERT = `${prefix}/REMOVE_ALERT`;

const ReducerRecord = Record({
  isAlert: false,
  alertMessage: "",
});

const Handler = {
  [SHOW_ALERT]: (state, payload) =>
    state.set("isAlert", true).set("alertMessage", payload),
  [REMOVE_ALERT]: (state) => state.set("isAlert", false),
  DEFAULT: (state) => state,
};

export default function alertReducer(
  state = new ReducerRecord(),
  { type, payload }
) {
  const handle = Handler[type] ?? Handler.DEFAULT;
  return handle(state, payload);
}

/**
 * Selectors
 */

const alertSelector = (state) => {
  return [state[moduleName].isAlert, state[moduleName].alertMessage];
};

/**
 * Hooks
 */

export const useAlert = () => {
  return useSelector(alertSelector);
};

/**
 * Actions
 */

let lastTimeout = null;

export const showAlert = (message) => {
  return (dispatch) => {
    dispatch({
      type: SHOW_ALERT,
      payload: message,
    });

    if (lastTimeout) {
      clearTimeout(lastTimeout);
    }

    lastTimeout = setTimeout(() => {
      dispatch(removeAlert());
    }, 3000);
  };
};

export const removeAlert = (message) => ({
  type: REMOVE_ALERT,
});
