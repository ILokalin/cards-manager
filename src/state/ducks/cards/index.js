import { Record } from "immutable";
import { useSelector } from "react-redux";
import { sortAZ } from "utils/sort";
import { http, dataAdapter } from "utils/http";
import { URN_STRING } from "config";
import { showAlert } from "state/ducks/events";
import {
  getCardsState,
  navBackwardState,
  navForwardState,
  selectCardState,
  setSortState,
  startLoadingState,
} from "./cardsUtils";

export const moduleName = "cards";
const prefix = `${moduleName}`;

export const SELECT_CARD = `${prefix}/SELECT_CARD`;
export const NAV_FORWARD = `${prefix}/NAV_FORWARD`;
export const NAV_BACKWARD = `${prefix}/NAV_BACKWARD`;
export const START_LOADING = `${prefix}/START_LOADING`;
export const GET_CARDS = `${prefix}/GET_CARDS`;
export const SET_SORT = `${prefix}/SET_SORT`;

const ReducerRecord = Record({
  cards: [],
  isLoading: false,
  sort: {
    key: "id",
    direction: sortAZ,
    isAZDirection: true,
  },
});

const Handler = {
  [START_LOADING]: startLoadingState,
  [GET_CARDS]: getCardsState,
  [NAV_BACKWARD]: navBackwardState,
  [NAV_FORWARD]: navForwardState,
  [SELECT_CARD]: selectCardState,
  [SET_SORT]: setSortState,
  DEFAULT: (state) => state,
};

export default function reducer(
  state = new ReducerRecord(),
  { type, payload }
) {
  const handle = Handler[type] ?? Handler.DEFAULT;
  return handle(state, payload);
}

/**
 * Selectors
 */

const cardsSelector = (state) => state[moduleName].cards;
const sortDirectionSelector = (state) => state[moduleName].sort.isAZDirection;
const sortKeySelector = (state) => state[moduleName].sort.key;

/**
 * Hooks
 */

export const useCards = () => {
  return useSelector(cardsSelector);
};

export const useSortDirection = () => {
  return useSelector(sortDirectionSelector);
};

export const useSortKey = () => {
  return useSelector(sortKeySelector);
};

/**
 * Helpers
 */

const isActionSuccess = ({ isError, message }, dispatch) => {
  if (isError) {
    dispatch(showAlert(message));
    return false;
  }
  return true;
};

/**
 * Actions
 */

export const getCards = () => async (dispatch) => {
  const response = await http(URN_STRING);

  if (isActionSuccess(response, dispatch)) {
    dispatch({
      type: GET_CARDS,
      payload: dataAdapter(response.data),
    });
  }
};

export const selectCard = (id) => {
  return {
    type: SELECT_CARD,
    payload: id,
  };
};

export const forward = () => {
  return {
    type: NAV_FORWARD,
  };
};

export const backward = () => {
  return {
    type: NAV_BACKWARD,
  };
};

export const setSort = (key) => {
  return {
    type: SET_SORT,
    payload: key,
  };
};
