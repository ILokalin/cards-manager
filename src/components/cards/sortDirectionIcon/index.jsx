import { useSortDirection } from "state/ducks/cards";

const PRIMARY = "--bs-primary";
const SECONDARY = "--bs-secondary";

export const SortDirectionIcon = ({ isKey }) => {
  const styleMap = new Map([
    [false, SECONDARY],
    [true, isKey ? PRIMARY : SECONDARY],
  ]);
  const isAZDirection = useSortDirection();

  return (
    <svg
      version="1.1"
      xmlns="http://www.w3.org/2000/svg"
      width="25"
      height="25"
      viewBox="0 0 16 16"
    >
      <path
        fill={`var(${styleMap.get(!isAZDirection)})`}
        d="M11 7h-6l3-4z"
      ></path>
      <path
        fill={`var(${styleMap.get(!!isAZDirection)})`}
        d="M5 9h6l-3 4z"
      ></path>
    </svg>
  );
};
