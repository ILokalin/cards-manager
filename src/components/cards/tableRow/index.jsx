import { forwardRef } from "react";
import { selectCard } from "state/ducks/cards";
import { useDispatch } from "react-redux";

const style = {
  row: {
    cursor: "pointer",
  },
  rowClass: "bg-primary text-light",
};

export const TableRow = forwardRef(
  ({ id, firstName, lastName, isActive }, ref) => {
    const dispatch = useDispatch();
    const rowSelectHandler = () => {
      dispatch(selectCard(id));
    };

    return (
      <tr
        className={`${isActive ? style.rowClass : ""}`}
        style={style.row}
        onClick={rowSelectHandler}
        ref={ref}
      >
        <td className="p-2 border">{id}</td>
        <td className="p-2 border">{firstName}</td>
        <td className="p-2 border">{lastName}</td>
      </tr>
    );
  }
);
