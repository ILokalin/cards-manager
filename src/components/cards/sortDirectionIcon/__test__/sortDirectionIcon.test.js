import { render } from "@testing-library/react";
import { SortDirectionIcon } from "../index";
import { useSortDirection } from "state/ducks/cards";

jest.mock("state/ducks/cards", () => ({
  ...jest.requireActual("state/ducks/cards"),
  useSortDirection: jest.fn().mockReturnValueOnce(true),
}));

describe("Sort direction icon", () => {
  it("should render icon without props", () => {
    const { asFragment } = render(<SortDirectionIcon />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render icon with isKey true and AZ direction", () => {
    const { asFragment } = render(<SortDirectionIcon isKey={true} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render icon with isKey true and ZA direction", () => {
    useSortDirection.mockResolvedValue(false);
    const { asFragment } = render(<SortDirectionIcon isKey={true} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
