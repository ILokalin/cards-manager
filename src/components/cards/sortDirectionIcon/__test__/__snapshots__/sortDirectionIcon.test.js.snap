// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort direction icon should render icon with isKey true and AZ direction 1`] = `
<DocumentFragment>
  <svg
    height="25"
    version="1.1"
    viewBox="0 0 16 16"
    width="25"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M11 7h-6l3-4z"
      fill="var(--bs-primary)"
    />
    <path
      d="M5 9h6l-3 4z"
      fill="var(--bs-secondary)"
    />
  </svg>
</DocumentFragment>
`;

exports[`Sort direction icon should render icon with isKey true and ZA direction 1`] = `
<DocumentFragment>
  <svg
    height="25"
    version="1.1"
    viewBox="0 0 16 16"
    width="25"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M11 7h-6l3-4z"
      fill="var(--bs-secondary)"
    />
    <path
      d="M5 9h6l-3 4z"
      fill="var(--bs-primary)"
    />
  </svg>
</DocumentFragment>
`;

exports[`Sort direction icon should render icon with sKey true 1`] = `
<DocumentFragment>
  <svg
    height="25"
    version="1.1"
    viewBox="0 0 16 16"
    width="25"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M11 7h-6l3-4z"
      fill="var(--bs-primary)"
    />
    <path
      d="M5 9h6l-3 4z"
      fill="var(--bs-secondary)"
    />
  </svg>
</DocumentFragment>
`;

exports[`Sort direction icon should render icon with sKey true and sortRevirse 1`] = `
<DocumentFragment>
  <svg
    height="25"
    version="1.1"
    viewBox="0 0 16 16"
    width="25"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M11 7h-6l3-4z"
      fill="var(--bs-secondary)"
    />
    <path
      d="M5 9h6l-3 4z"
      fill="var(--bs-primary)"
    />
  </svg>
</DocumentFragment>
`;

exports[`Sort direction icon should render icon without props 1`] = `
<DocumentFragment>
  <svg
    height="25"
    version="1.1"
    viewBox="0 0 16 16"
    width="25"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M11 7h-6l3-4z"
      fill="var(--bs-secondary)"
    />
    <path
      d="M5 9h6l-3 4z"
      fill="var(--bs-secondary)"
    />
  </svg>
</DocumentFragment>
`;
